File: src/main/java/org/github/cuipengfei/BufferedImages2Tiff.java
package org.github.cuipengfei;

import com.twelvemonkeys.imageio.plugins.tiff.TIFFImageWriter;
import lombok.extern.slf4j.Slf4j;

import javax.imageio.IIOImage;
import javax.imageio.ImageWriteParam;
import javax.imageio.ImageWriter;
import javax.imageio.stream.ImageOutputStream;
import java.awt.image.BufferedImage;
import java.io.IOException;
import java.io.OutputStream;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import static javax.imageio.ImageIO.createImageOutputStream;
import static javax.imageio.ImageIO.getImageWritersByFormatName;

@Slf4j
public class BufferedImages2Tiff {

    /**
     * Convert from a list of buffered images to tiff output stream
     *
     * @param bufferedImages a list of buffered images
     * @param output         tiff output stream
     * @param compression    compression type, such as "LZW", "JPEG", "PackBits", "Deflate"
     * @throws IOException            if an I/O error occurs
     * @throws ClassNotFoundException if the class cannot be located
     */
    public void bufferedImages2TiffOutputStream(
            List<BufferedImage> bufferedImages, OutputStream output, String compression)
            throws IOException, ClassNotFoundException {

        Optional<ImageWriter> tiffWriterOptional = findTiffWriter();

        if (tiffWriterOptional.isPresent()) {
            ImageWriter tiffWriter = tiffWriterOptional.get();
            ImageWriteParam writeParam = configWriteParam(tiffWriter, compression);

            try (ImageOutputStream imageOutputStream = createImageOutputStream(output)) {
                tiffWriter.setOutput(imageOutputStream);
                tiffWriter.prepareWriteSequence(null);

                for (BufferedImage bufferedImage : bufferedImages) {
                    tiffWriter.writeToSequence(new IIOImage(bufferedImage, null, null), writeParam);
                }

                tiffWriter.endWriteSequence();
            }
            tiffWriter.dispose();
            log.info("Buffered images to tiff output stream, number of pages: {}", bufferedImages.size());
        } else {
            throw new ClassNotFoundException("Can not find tiff writer class");
        }
    }

    private static ImageWriteParam configWriteParam(ImageWriter tiffWriter, String compression) {
        ImageWriteParam writeParam = tiffWriter.getDefaultWriteParam();
        writeParam.setCompressionMode(ImageWriteParam.MODE_EXPLICIT);
        writeParam.setCompressionType(compression);
        return writeParam;
    }

    private static Optional<ImageWriter> findTiffWriter() {
        List<ImageWriter> actualList = new ArrayList<>();
        getImageWritersByFormatName("TIFF").forEachRemaining(actualList::add);
        log.info("Find tiff writer class: {}", actualList);
        return actualList.stream().filter(writer -> writer instanceof TIFFImageWriter).findFirst();
    }
}

File: src/main/kotlin/com/github/pdf2tiff/BufferedImages2Tiff.kt
package com.github.pdf2tiff

import com.twelvemonkeys.imageio.plugins.tiff.TIFFImageWriter
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.awt.image.BufferedImage
import java.io.IOException
import java.io.OutputStream
import java.util.*
import javax.imageio.IIOImage
import javax.imageio.ImageIO
import javax.imageio.ImageWriteParam
import javax.imageio.ImageWriter

class BufferedImages2Tiff {
    private val log: Logger = LoggerFactory.getLogger(this.javaClass)

    /**
     * Convert from a list of buffered images to tiff output stream
     *
     * @param bufferedImages a list of buffered images
     * @param output         tiff output stream
     * @param compression    compression type, such as "LZW", "JPEG", "PackBits", "Deflate"
     * @throws IOException            if an I/O error occurs
     * @throws ClassNotFoundException if the class cannot be located
     */
    @Throws(IOException::class, ClassNotFoundException::class)
    fun bufferedImages2TiffOutputStream(
        bufferedImages: List<BufferedImage>,
        output: OutputStream,
        compression: String
    ) {
        val tiffWriterOptional = findTiffWriter()

        if (tiffWriterOptional.isPresent) {
            ImageIO.createImageOutputStream(output).use { imageOutputStream ->
                val tiffWriter = tiffWriterOptional.get()
                val writeParam = configWriteParam(tiffWriter, compression)

                tiffWriter.output = imageOutputStream
                tiffWriter.prepareWriteSequence(null)

                for (bufferedImage in bufferedImages) {
                    tiffWriter.writeToSequence(IIOImage(bufferedImage, null, null), writeParam)
                }

                tiffWriter.endWriteSequence()
                tiffWriter.dispose()
            }

            log.info("Buffered images to tiff output stream, number of pages: ${bufferedImages.size}")
        } else {
            throw ClassNotFoundException("Can not find tiff writer class")
        }
    }

    private fun configWriteParam(tiffWriter: ImageWriter, compression: String): ImageWriteParam {
        val writeParam = tiffWriter.defaultWriteParam
        writeParam.compressionMode = ImageWriteParam.MODE_EXPLICIT
        writeParam.compressionType = compression
        return writeParam
    }

    private fun findTiffWriter(): Optional<ImageWriter> {
        val tiffWriters = ImageIO.getImageWritersByFormatName("TIFF").asSequence()
        val first = tiffWriters.filterIsInstance<TIFFImageWriter>().firstOrNull()
        log.info("Try to find tiff writer class: $first")
        return first.toOptional()
    }

    private fun <T : Any> T?.toOptional(): Optional<T> = Optional.ofNullable(this)
}